File: apps/android/app/src/main/java/com/rayact/engine/RayactEngineSession.kt
package com.rayact.engine

import android.view.Surface
import com.rayact.app.RayactRenderScheduler

class RayactEngineSession private constructor(val nativeHandle: Long) {
    val host = RayactHost(this, RayactRenderScheduler(this))

    // Cleared on destroy so callers (e.g. the shared DevClientBridge) can avoid
    // invoking native methods on a torn-down engine handle.
    @Volatile private var destroyed = false
    fun isAlive(): Boolean = !destroyed

    fun acquireGraphics(): Boolean = nativeAcquireGraphics(nativeHandle)
    fun releaseGraphics() {
        host.renderScheduler.stopRendering()
        nativeReleaseGraphics(nativeHandle)
    }
    fun destroy() {
        destroyed = true
        nativeDestroy(nativeHandle)
    }

    fun loadSource(source: String): Boolean = nativeLoadScript(nativeHandle, 0, source)
    fun loadDevClient(source: String): Boolean = nativeLoadScript(nativeHandle, 0, source)
    fun loadDevServer(url: String): Boolean = nativeLoadScript(nativeHandle, 1, url)
    fun loadBytecode(bytes: ByteArray): Boolean = nativeLoadBytecode(nativeHandle, bytes)
    private fun devTransportScript(serverUrl: String): String = """
        globalThis.__RAYACT_DEV_SERVER__ = ${org.json.JSONObject.quote(serverUrl)};
        globalThis.__rayactDevFetch = function(url) { return rayactDevFetch(url); };
    """.trimIndent()

    /**
     * Load the dev bootstrap with its transport globals in a SINGLE script.
     * nativeLoadScript keeps only one pending script slot, so a separate
     * injectDevTransport() call would be clobbered by the bootstrap load and the
     * bootstrap would run without __RAYACT_DEV_SERVER__ (→ black screen).
     */
    fun loadDevBootstrap(serverUrl: String, source: String): Boolean =
        loadSource(devTransportScript(serverUrl) + "\n" + source)

    fun applyModuleUpdate(path: String, source: String): Boolean =
        nativeApplyModuleUpdate(nativeHandle, path, source)

    fun createSurface(surface: Surface, density: Float): Int =
        nativeCreateSurface(nativeHandle, surface, density)

    fun resizeSurface(surfaceId: Int, width: Int, height: Int, density: Float) =
        nativeResizeSurface(nativeHandle, surfaceId, width, height, density)

    fun relayoutOnSurfaceResizeEnabled(): Boolean =
        nativeRelayoutOnSurfaceResizeEnabled(nativeHandle)

    fun setSafeAreaInsets(top: Float, right: Float, bottom: Float, left: Float) =
        nativeSetSafeAreaInsets(nativeHandle, top, right, bottom, left)

    fun setKeyboardInsets(heightDp: Float, visible: Boolean, durationMs: Float) =
        nativeSetKeyboardInsets(nativeHandle, heightDp, visible, durationMs)

    fun destroySurface(surfaceId: Int) = nativeDestroySurface(nativeHandle, surfaceId)
    fun pushSurface(surfaceId: Int) = nativePushSurface(nativeHandle, surfaceId)
    fun popSurface(): Int = nativePopSurface(nativeHandle)
    fun getFocusedSurfaceId(): Int = nativeGetFocusedSurfaceId(nativeHandle)

    fun nativeRenderFrame(frameTimeNanos: Long, deltaNanos: Long): Boolean =
        nativeRenderFrame(nativeHandle, frameTimeNanos, deltaNanos)

    fun nativeNextJSTimerDelayMs(): Float = nativeNextJSTimerDelayMs(nativeHandle)
    fun nativeToggleDevMenu() = nativeToggleDevMenu(nativeHandle)
    fun nativeOnBackPressed() = nativeOnBackPressed(nativeHandle)

    fun nativeTouch(action: Int, id: Int, x: Float, y: Float) =
        nativeTouch(nativeHandle, action, id, x, y)

    fun nativeSetTextInputContent(
        nodeId: Int, text: String,
        selectionStart: Int, selectionEnd: Int,
        composingStart: Int, composingEnd: Int
    ) = nativeSetTextInputContent(
        nativeHandle, nodeId, text, selectionStart, selectionEnd, composingStart, composingEnd
    )

    fun nativeBlurTextInput() = nativeBlurTextInput(nativeHandle)
    fun nativeImeHiddenBySystem() = nativeImeHiddenBySystem(nativeHandle)

    fun nativePushExternalViewFrame(nodeId: Int, buffer: android.hardware.HardwareBuffer) =
        nativePushExternalViewFrame(nativeHandle, nodeId, buffer)

    fun nativeExternalViewTextChanged(nodeId: Int, text: String) =
        nativeExternalViewTextChanged(nativeHandle, nodeId, text)

    fun nativeSetExternalViewInsets(nodeId: Int, l: Float, t: Float, r: Float, b: Float) =
        nativeSetExternalViewInsets(nativeHandle, nodeId, l, t, r, b)

    companion object {
        init { System.loadLibrary("rayact") }

        fun create(dataPath: String): RayactEngineSession? {
            val handle = nativeCreate(dataPath)
            if (handle == 0L) return null
            val session = RayactEngineSession(handle)
            nativeRegisterHost(handle, session.host.hostCallbacks)
            RayactPlatformViews.bindSession(session)
            return session
        }

        const val TOUCH_DOWN = 0
        const val TOUCH_UP = 1
        const val TOUCH_MOVE = 2

        @JvmStatic external fun nativeCreate(dataPath: String): Long
        @JvmStatic external fun nativeRegisterHost(handle: Long, callbacks: RayactEngineHostCallbacks)
        @JvmStatic external fun nativeDestroy(handle: Long)
        @JvmStatic external fun nativeAcquireGraphics(handle: Long): Boolean
        @JvmStatic external fun nativeReleaseGraphics(handle: Long)
        @JvmStatic external fun nativeLoadScript(handle: Long, mode: Int, arg: String): Boolean
        @JvmStatic external fun nativeLoadBytecode(handle: Long, bytes: ByteArray): Boolean
        @JvmStatic external fun nativeApplyModuleUpdate(handle: Long, path: String, source: String): Boolean
        @JvmStatic external fun nativeToggleDevMenu(handle: Long)
        @JvmStatic external fun nativeCreateSurface(handle: Long, surface: Surface, density: Float): Int
        @JvmStatic external fun nativeResizeSurface(handle: Long, surfaceId: Int, width: Int, height: Int, density: Float)
        @JvmStatic external fun nativeRelayoutOnSurfaceResizeEnabled(handle: Long): Boolean
        @JvmStatic external fun nativeSetSafeAreaInsets(handle: Long, top: Float, right: Float, bottom: Float, left: Float)
        @JvmStatic external fun nativeSetKeyboardInsets(handle: Long, heightDp: Float, visible: Boolean, durationMs: Float)
        @JvmStatic external fun nativeDestroySurface(handle: Long, surfaceId: Int)
        @JvmStatic external fun nativePushSurface(handle: Long, surfaceId: Int)
        @JvmStatic external fun nativePopSurface(handle: Long): Int
        @JvmStatic external fun nativeGetFocusedSurfaceId(handle: Long): Int
        @JvmStatic external fun nativeRenderFrame(handle: Long, frameTimeNanos: Long, deltaNanos: Long): Boolean
        @JvmStatic external fun nativeNextJSTimerDelayMs(handle: Long): Float
        @JvmStatic external fun nativeTouch(handle: Long, action: Int, id: Int, x: Float, y: Float)
        @JvmStatic external fun nativeOnBackPressed(handle: Long)
        @JvmStatic external fun nativeSetTextInputContent(
            handle: Long, nodeId: Int, text: String,
            selectionStart: Int, selectionEnd: Int, composingStart: Int, composingEnd: Int
        )
        @JvmStatic external fun nativeBlurTextInput(handle: Long)
        @JvmStatic external fun nativeImeHiddenBySystem(handle: Long)
        @JvmStatic external fun nativePushExternalViewFrame(handle: Long, nodeId: Int, buffer: android.hardware.HardwareBuffer)
        @JvmStatic external fun nativeExternalViewTextChanged(handle: Long, nodeId: Int, text: String)
        @JvmStatic external fun nativeSetExternalViewInsets(handle: Long, nodeId: Int, l: Float, t: Float, r: Float, b: Float)
    }
}

File: apps/ios/RayactEngineSession.swift
import Foundation
import UIKit

final class RayactEngineSession {
    static let TOUCH_DOWN = 0
    static let TOUCH_UP = 1
    static let TOUCH_MOVE = 2

    let nativeHandle: RayactIOSHandle
    private(set) lazy var host: RayactHost = RayactHost(
        session: self,
        renderScheduler: RayactRenderScheduler(session: self)
    )
    private var hostBridge: RayactHostCallbacksBridge?
    private var destroyed = false

    private init(handle: RayactIOSHandle) {
        nativeHandle = handle
    }

    func isAlive() -> Bool { !destroyed }

    func acquireGraphics() -> Bool {
        RayactNativeBridge.sessionAcquireGraphics(nativeHandle)
    }

    func releaseGraphics() {
        host.renderScheduler.stopRendering()
        RayactNativeBridge.sessionReleaseGraphics(nativeHandle)
    }

    func destroy() {
        destroyed = true
        hostBridge = nil
        RayactNativeBridge.sessionDestroy(nativeHandle)
    }

    func loadSource(_ source: String) -> Bool {
        source.withCString { RayactNativeBridge.sessionLoadScript(nativeHandle, 0, $0) }
    }

    func loadDevClient(_ source: String) -> Bool {
        loadSource(source)
    }

    func loadDevServer(_ url: String) -> Bool {
        url.withCString { RayactNativeBridge.sessionLoadScript(nativeHandle, 1, $0) }
    }

    func loadDevBootstrap(serverUrl: String, source: String) -> Bool {
        let transport = """
        globalThis.__RAYACT_DEV_SERVER__ = \(Self.jsonQuote(serverUrl));
        globalThis.__rayactDevFetch = function(url) { return rayactDevFetch(url); };
        """
        return loadSource(transport + "\n" + source)
    }

    func loadBytecode(_ bytes: Data) -> Bool {
        bytes.withUnsafeBytes { buffer in
            guard let base = buffer.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return false }
            return RayactNativeBridge.sessionLoadBytecode(nativeHandle, base, Int32(bytes.count))
        }
    }

    func applyModuleUpdate(path: String, source: String) -> Bool {
        path.withCString { pathPtr in
            source.withCString { sourcePtr in
                RayactNativeBridge.sessionApplyModuleUpdate(nativeHandle, pathPtr, sourcePtr)
            }
        }
    }

    func createSurface(metalLayer: UnsafeMutableRawPointer, density: Float, widthPx: Int, heightPx: Int, scale: Float) -> Int {
        Int(RayactNativeBridge.sessionCreateSurface(
            nativeHandle,
            metalLayer,
            density,
            Int32(widthPx),
            Int32(heightPx),
            scale
        ))
    }

    func resizeSurface(surfaceId: Int, width: Int, height: Int, density: Float) {
        RayactNativeBridge.sessionResizeSurface(
            nativeHandle,
            Int32(surfaceId),
            Int32(width),
            Int32(height),
            density
        )
    }

    func relayoutOnSurfaceResizeEnabled() -> Bool {
        RayactNativeBridge.sessionRelayoutOnSurfaceResizeEnabled(nativeHandle)
    }

    func setSafeAreaInsets(top: Float, right: Float, bottom: Float, left: Float) {
        RayactNativeBridge.sessionSetSafeAreaInsets(nativeHandle, top, right, bottom, left)
    }

    func setKeyboardInsets(heightDp: Float, visible: Bool, durationMs: Float) {
        RayactNativeBridge.sessionSetKeyboardInsets(nativeHandle, heightDp, visible, durationMs)
    }

    func keyEvent(type: Int32, key: String, code: String, text: String = "", modifiers: UIKeyModifierFlags = []) {
        key.withCString { keyPtr in
            code.withCString { codePtr in
                text.withCString { textPtr in
                    RayactNativeBridge.sessionKeyEvent(
                        nativeHandle, type, keyPtr, codePtr, textPtr, false,
                        modifiers.contains(.control), modifiers.contains(.alternate),
                        modifiers.contains(.shift), modifiers.contains(.command)
                    )
                }
            }
        }
    }

    func destroySurface(_ surfaceId: Int) {
        RayactNativeBridge.sessionDestroySurface(nativeHandle, Int32(surfaceId))
    }

    func pushSurface(_ surfaceId: Int) {
        RayactNativeBridge.sessionPushSurface(nativeHandle, Int32(surfaceId))
    }

    func popSurface() -> Int {
        Int(RayactNativeBridge.sessionPopSurface(nativeHandle))
    }

    func getFocusedSurfaceId() -> Int {
        Int(RayactNativeBridge.sessionGetFocusedSurfaceId(nativeHandle))
    }

    func nativeRenderFrame(frameTimeNanos: Int64, deltaNanos: Int64) -> Bool {
        RayactNativeBridge.sessionRenderFrame(nativeHandle, frameTimeNanos, deltaNanos)
    }

    func nativeNextJSTimerDelayMs() -> Float {
        RayactNativeBridge.sessionNextJSTimerDelayMs(nativeHandle)
    }

    func nativeToggleDevMenu() {
        RayactNativeBridge.sessionToggleDevMenu(nativeHandle)
    }

    func nativeOnBackPressed() {
        RayactNativeBridge.sessionOnBackPressed(nativeHandle)
    }

    func refreshAppearance() {
        RayactNativeBridge.sessionRefreshAppearance(nativeHandle)
    }

    func nativeTouch(action: Int, id: Int, x: Float, y: Float) {
        RayactNativeBridge.sessionTouch(nativeHandle, Int32(action), Int32(id), x, y)
    }

    func nativeSetTextInputContent(
        nodeId: Int,
        text: String,
        selectionStart: Int,
        selectionEnd: Int,
        composingStart: Int,
        composingEnd: Int
    ) {
        text.withCString { textPtr in
            RayactNativeBridge.sessionSetTextInputContent(
                nativeHandle,
                Int32(nodeId),
                textPtr,
                Int32(selectionStart),
                Int32(selectionEnd),
                Int32(composingStart),
                Int32(composingEnd)
            )
        }
    }

    func nativeBlurTextInput() {
        RayactNativeBridge.sessionBlurTextInput(nativeHandle)
    }

    func nativeImeHiddenBySystem() {
        RayactNativeBridge.sessionImeHiddenBySystem(nativeHandle)
    }

    static func create(dataPath: String) -> RayactEngineSession? {
        let handle = dataPath.withCString { RayactNativeBridge.sessionCreate($0) }
        guard handle != 0 else { return nil }
        let session = RayactEngineSession(handle: handle)
        _ = session.host
        let bridge = RayactHostCallbacksBridge(host: session.host)
        session.hostBridge = bridge
        RayactNativeBridge.sessionRegisterHost(handle, bridge.callbacksPointer)
        return session
    }

    private static func jsonQuote(_ value: String) -> String {
        if let data = try? JSONEncoder().encode(value),
           let quoted = String(data: data, encoding: .utf8) {
            return quoted
        }
        return "\"\(value.replacingOccurrences(of: "\"", with: "\\\""))\""
    }
}
